langchain/sql_database.ts

```diff
@@ -0,0 +1,99 @@
+import { DataSource } from "typeorm";
+import {
+    generateTableInfoFromTables,
+    getTableAndColumnsName,
+    verifyIgnoreTablesExistInDatabase,
+    verifyIncludeTablesExistInDatabase,
+    verifyListTablesExistInDatabase
+} from './util/sql_utils';
+
+export interface SQLDatabaseParams {
+    appDataSource: DataSource;
+    includesTables?: Array<string>;
+    ignoreTables?: Array<string>;
+    sampleRowsInTableInfo?: number;
+}
+
+export interface SqlTable {
+    tableName: string;
+    columns: SqlColumn[];
+}
+
+export interface SqlColumn {
+    columnName: string;
+    dataType?: string;
+    isNullable?: boolean;
+}
+
+export class SQLDatabase {
+    appDataSource: DataSource;
+
+    allTables: Array<SqlTable> = [];
+
+    includesTables: Array<string> = [];
+
+    ignoreTables: Array<string> = [];
+
+    sampleRowsInTableInfo = 3;
+
+    constructor(fields: SQLDatabaseParams) {
+        this.appDataSource = fields.appDataSource;
+        if (fields?.includesTables && fields?.ignoreTables) {
+            throw new Error(
+                "Cannot specify both include_tables and ignoreTables"
+            );
+
+        }
+        this.includesTables = fields?.includesTables ?? [];
+        this.ignoreTables = fields?.ignoreTables ?? [];
+        this.sampleRowsInTableInfo = fields?.sampleRowsInTableInfo ?? this.sampleRowsInTableInfo;
+    }
+
+    async fromDataSource(fields: SQLDatabaseParams): Promise<SQLDatabase> {
+        const sqlDatabase = new SQLDatabase(fields);
+        sqlDatabase.allTables = await getTableAndColumnsName(fields.appDataSource);
+        verifyIncludeTablesExistInDatabase(this.allTables, this.includesTables);
+        verifyIgnoreTablesExistInDatabase(this.allTables, this.ignoreTables);
+
+        return sqlDatabase;
+    }
+
+    /**
+     * """Get information about specified tables.
+     *
+     * Follows best practices as specified in: Rajkumar et al, 2022
+     * (https://arxiv.org/abs/2204.00498)
+     *
+     * If `sample_rows_in_table_info`, the specified number of sample rows will be
+     * appended to each table description. This can increase performance as
+     * demonstrated in the paper.
+     */
+    async getTableInfo(targetTables?: Array<string>): Promise<string> {
+        let selectedTables = this.allTables;
+        if (targetTables && targetTables.length > 0) {
+            verifyListTablesExistInDatabase(this.allTables, targetTables, 'Wrong target table name:');
+            selectedTables = this.allTables.filter((currentTable) => targetTables.includes(currentTable.tableName));
+        }
+
+        return generateTableInfoFromTables(selectedTables, this.appDataSource, this.sampleRowsInTableInfo);
```

Comment:
I chose to put most of the function outside the class to make to uncoupled it a bit.